###    List e Dict Comprehensions, Comprehensions aninhadas, Generator Expressions, Funções Pré-Definidas, Numéricas e Lógicas que trabalham com iteradores.

LingProg - Lukas Barbosa de Oliveira - RA: 81714448 <br><br>

<b>1 Faça um programa que peça uma nota, entre zero e dez. Mostre uma mensagem
caso o valor seja inválido e continue pedindo até que o usuário informe um valor
válido.</b>

In [4]:
#Maybe

def maybe(fnc):
    def inner(*args):
        for a in args:
            if isinstance(a, Exception):
                return a
        try:
            return fnc(*args)
        except Exception as e:
            return e
    return inner

In [5]:
def maybe_number(fnc):
    def inner(args):
        try:
            return fnc(args)
        except ValueError as e:
            print('Nota inválida, digite apenas uma nota de 0 a 10.')
            return inner(args)
    return inner

get_numero = maybe_number(lambda mensagem: int(input(mensagem)))

def maybe_entre_numeros(inicio, fim):
    def maybe(fnc):
        def inner():
            num = get_numero(f'Digite uma nota entre {inicio} e {fim}: ')
            if (num >= inicio and num <= fim):
                return fnc(num)
            else:
                print('Nota inválida, digite apenas uma nota de 0 a 10.')
                return inner()
        return inner
    return maybe

mostrar_nota = lambda: maybe_entre_numeros(0, 10)(lambda nota: print('A nota digitada foi', nota))()
    
mostrar_nota()

Digite uma nota entre 0 e 10: 14
Nota inválida, digite apenas uma nota de 0 a 10.
Digite uma nota entre 0 e 10: 8
A nota digitada foi 8


<b>2 Faça um programa que leia e valide as seguintes informações:</b><br><br>
Nome: maior que 3 caracteres;<br>
Idade: entre 0 e 150;<br>
Salário: maior que zero;<br>
Sexo: 'f' ou 'm';<br>
Estado Civil: 's', 'c', 'v', 'd';

In [11]:
nome = ''
idade = 0
salario = 0
sexo = ''
estado_civil = ''

while len(nome)<=3:
    nome = input('Nome: (com mais de 3 letras)')

while idade < 0 or idade > 150 :
    idade = input('Idade:(entre 0 e 150) ')

while salario <= 0:
    salario = float(input('Salário: (maior que 0) '))

while sexo != 'f'  and sexo != 'm':
    sexo = input('Sexo: [m/f]')

while estado_civil != 's' and estado_civil != 'c' and estado_civil != 'v' and estado_civil != 'd':
    estado_civil = input('Estado Civil: [s/c/v/d]')

Nome: (com mais de 3 letras)lukas
Salário: (maior que 0) 200
Sexo: [m/f]g
Sexo: [m/f]m
Estado Civil: [s/c/v/d]t
Estado Civil: [s/c/v/d]s


<b>Faça um programa que leia 5 números e informe o maior número.<b>

In [6]:
numeros = lambda: [get_numero(f"Digite o {x}º numero: ") for x in range(1, 6)]

print(f'\nO maior numero é {max(numeros())}')

Digite o 1º numero: 17
Digite o 2º numero: 26
Digite o 3º numero: 34
Digite o 4º numero: 2
Digite o 5º numero: 29

O maior numero é 34


<b>4 Faça um programa que leia 5 números e informe a soma e a média dos números.</b>


In [8]:
(lambda x: print(f'----------\nSoma: {sum(x)} \nMedia: {sum(x)/len(x)}'))(numeros())

Digite o 1º numero: 5
Digite o 2º numero: 6
Digite o 3º numero: 2
Digite o 4º numero: 3
Digite o 5º numero: 12
----------
Soma: 28 
Media: 5.6


<b>5 Faça um programa que imprima na tela apenas os números ímpares entre 1 e 50</b>

In [9]:
print(f'Numeros ímpares entre 1 e 50: {(lambda: [x for x in range(1,51) if not(x % 2 == 0)])()}')

Numeros ímpares entre 1 e 50: [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49]


<b>6 Desenvolva um gerador de tabuada, capaz de gerar a tabuada de qualquer número
inteiro entre 1 a 10. O usuário deve informar de qual numero ele deseja ver a
tabuada. A saída deve ser conforme o exemplo abaixo:</b><br><br>
Tabuada de 5:<br>
5 X 1 = 5<br>
5 X 2 = 10<br>
...<br>
5 X 10 = 50<br>

In [10]:
for i, x in enumerate((lambda: maybe_entre_numeros(0, 10)(lambda x: [x * i for i in range(1, 11)])())()):
    if i == 0:
        print(f'Tabuada de {x}: ')
        y = x
    print(f'{y} X {i} = {x}')

Digite uma nota entre 0 e 10: 7
Tabuada de 7: 
7 X 0 = 7
7 X 1 = 14
7 X 2 = 21
7 X 3 = 28
7 X 4 = 35
7 X 5 = 42
7 X 6 = 49
7 X 7 = 56
7 X 8 = 63
7 X 9 = 70


<b>7 A série de Fibonacci é formada pela seqüência 1,1,2,3,5,8,13,21,34,55,... Faça um
programa capaz de gerar a série até o n−ésimo termo.</b>

In [11]:
n = int(input('Digite até qual termo a série será gerada: '))

a, b = 0, 1 
fibonacci = []

for x in range(n):
    fibonacci.append(b)
    a, b = b, a+b
    
print(fibonacci)

Digite até qual termo a série será gerada: 9
[1, 1, 2, 3, 5, 8, 13, 21, 34]


<b>8 Faça um programa que calcule o fatorial de um número inteiro fornecido pelo
usuário. Ex.: 5!=5.4.3.2.1=120</b>

In [12]:
from functools import reduce

fatorial = lambda numero: reduce(lambda a, b: a * b, [i for i in range(1, numero+1)])
x = get_numero("Digite o numero que deseja calcular o fatorial")
print(f'{x}! = {fatorial(x)}')

Digite o numero que deseja calcular o fatorial7
7! = 5040


- <b>9 O Sr. Manoel Joaquim possui uma grande loja de artigos de RS 1,99, com cerca de
10 caixas. Para agilizar o cálculo de quanto cada cliente deve pagar ele desenvolveu
um tabela que contém o número de itens que o cliente comprou e ao lado o valor da
conta. Desta forma a atendente do caixa precisa apenas contar quantos itens o
cliente está levando e olhar na tabela de preços. Você foi contratado para
desenvolver o programa que monta esta tabela de preços, que conterá os preços de
1 até 50 produtos, conforme o exemplo abaixo:</b><br><br>
Lojas Quase Dois - Tabela de preços<br>
1 - RS 1.99<br>
2 - RS 3.98<br>
...<br>
50 - R$ 99.50<br>

In [13]:
print('Lojas Quase Dois - Tabela de preços')
[print(f'{i} - R$ {1.99 * i:.2f}') for i in range(1, 51)]
print()

Lojas Quase Dois - Tabela de preços
1 - R$ 1.99
2 - R$ 3.98
3 - R$ 5.97
4 - R$ 7.96
5 - R$ 9.95
6 - R$ 11.94
7 - R$ 13.93
8 - R$ 15.92
9 - R$ 17.91
10 - R$ 19.90
11 - R$ 21.89
12 - R$ 23.88
13 - R$ 25.87
14 - R$ 27.86
15 - R$ 29.85
16 - R$ 31.84
17 - R$ 33.83
18 - R$ 35.82
19 - R$ 37.81
20 - R$ 39.80
21 - R$ 41.79
22 - R$ 43.78
23 - R$ 45.77
24 - R$ 47.76
25 - R$ 49.75
26 - R$ 51.74
27 - R$ 53.73
28 - R$ 55.72
29 - R$ 57.71
30 - R$ 59.70
31 - R$ 61.69
32 - R$ 63.68
33 - R$ 65.67
34 - R$ 67.66
35 - R$ 69.65
36 - R$ 71.64
37 - R$ 73.63
38 - R$ 75.62
39 - R$ 77.61
40 - R$ 79.60
41 - R$ 81.59
42 - R$ 83.58
43 - R$ 85.57
44 - R$ 87.56
45 - R$ 89.55
46 - R$ 91.54
47 - R$ 93.53
48 - R$ 95.52
49 - R$ 97.51
50 - R$ 99.50



<b>10 O Sr. Manoel Joaquim acaba de adquirir uma panificadora e pretende implantar a
metodologia da tabelinha, que já é um sucesso na sua loja de 1,99. Você foi
contratado para desenvolver o programa que monta a tabela de preços de pães, de
1 até 50 pães, a partir do preço do pão informado pelo usuário, conforme o exemplo
abaixo:</b><br><br>
Preço do pão: RS 0.18<br>
Panificadora Pão de Ontem - Tabela de preços<br>
1 - RS 0.18<br>
2 - RS 0.36<br>
...<br>
50 - RS 9.00

In [15]:
get_numero_float = maybe_number(lambda mensagem: float(input(mensagem)))
y = get_numero_float('Preço do pão: R$')
print('Panificadora Pão de Ontem - Tabela de preços')
[print(f'{i} - R$ {y * i:.2f}') for i in range(1, 51)]
print()

Preço do pão: R$0.20
Panificadora Pão de Ontem - Tabela de preços
1 - R$ 0.20
2 - R$ 0.40
3 - R$ 0.60
4 - R$ 0.80
5 - R$ 1.00
6 - R$ 1.20
7 - R$ 1.40
8 - R$ 1.60
9 - R$ 1.80
10 - R$ 2.00
11 - R$ 2.20
12 - R$ 2.40
13 - R$ 2.60
14 - R$ 2.80
15 - R$ 3.00
16 - R$ 3.20
17 - R$ 3.40
18 - R$ 3.60
19 - R$ 3.80
20 - R$ 4.00
21 - R$ 4.20
22 - R$ 4.40
23 - R$ 4.60
24 - R$ 4.80
25 - R$ 5.00
26 - R$ 5.20
27 - R$ 5.40
28 - R$ 5.60
29 - R$ 5.80
30 - R$ 6.00
31 - R$ 6.20
32 - R$ 6.40
33 - R$ 6.60
34 - R$ 6.80
35 - R$ 7.00
36 - R$ 7.20
37 - R$ 7.40
38 - R$ 7.60
39 - R$ 7.80
40 - R$ 8.00
41 - R$ 8.20
42 - R$ 8.40
43 - R$ 8.60
44 - R$ 8.80
45 - R$ 9.00
46 - R$ 9.20
47 - R$ 9.40
48 - R$ 9.60
49 - R$ 9.80
50 - R$ 10.00



<b>11 O Sr. Manoel Joaquim expandiu seus negócios para além dos negócios de 1,99 e
agora possui uma loja de conveniências. Faça um programa que implemente uma
caixa registradora rudimentar. O programa deverá receber um número
desconhecido de valores referentes aos preços das mercadorias. Um valor zero deve
ser informado pelo operador para indicar o final da compra. O programa deve então
mostrar o total da compra e perguntar o valor em dinheiro que o cliente forneceu, 
para então calcular e mostrar o valor do troco. Após esta operação, o programa
deverá voltar ao ponto inicial, para registrar a próxima compra. A saída deve ser
conforme o exemplo abaixo:</b><br><br>
Lojas Tabajara<br>
Produto 1: RS 2.20<br>
Produto 2: RS 5.80<br>
Produto 3: RS 0<br>
Total: RS 9.00<br>
Dinheiro: RS 20.00<br>
Troco: RS 11.00<br>
...

In [16]:
print('Lojas Tabajara')
valor = -1
total = 0
n = 0
while valor != 0:
    n += 1
    valor = float(input(f'Produto {n}: R$ '))
    total += valor                  
print(f'Total: R$ {total:.2f}')
dinheiro = float(input('Dinheiro: R$ '))
print(f'Troco: R$ {dinheiro - total}\n...')

Lojas Tabajara
Produto 1: R$ 4
Produto 2: R$ 1
Produto 3: R$ 3
Produto 4: R$ 5
Produto 5: R$ 2
Produto 6: R$ 3
Produto 7: R$ 0
Total: R$ 18.00
Dinheiro: R$ 27
Troco: R$ 9.0
...


<b>12 Foi feita uma estatística em cinco cidades brasileiras para coletar dados sobre
acidentes de trânsito. Foram obtidos os seguintes dados:<.b><br><br>
- Código da cidade;
- Número de veículos de passeio (em 1999);
- Número de acidentes de trânsito com vítimas (em 1999). Deseja-se saber:
- Qual o maior e menor índice de acidentes de transito e a que cidade pertence;
- Qual a média de veículos nas cinco cidades juntas;
- Qual a média de acidentes de trânsito nas cidades com menos de 2.000 veículos de
passeio.

In [ ]:
maior = menor = count = soma_veiculos = soma_acidentes = soma_2k = 0
cid_maior = cid_menor = ''

for c in range(1,6):
    cidade      = input("\nDigite o nome da cidade.....................: ")
    codigo      = int(input("Digite o código da cidade...................: "))
    veiculos    = int(input("Numero de veiculos de passeio...............: "))
    acidentes   = int(input("Numero de acidentes de transito com vitimas.: "))

    soma_veiculos += veiculos
    soma_acidentes += acidentes

    if acidentes > maior:
        maior = acidentes
        cid_maior = cidade

    if acidentes < menor or c == 1:
        menor = acidentes
        cid_menor = cidade

    if veiculos < 2000:
        soma_2k += acidentes
        count += 1
media_nas_5_cidades = soma_veiculos / c
media_2k = soma_2k / count


print("\n"+"-="*30 + '-')

print(
    f"O menor indice de acidentes de transito {menor} cidade que pertence {cid_menor}",
    f"O maior indice de acidenstes de transito {maior} cidade que pertence {cid_maior}",
    f"Media de veiculos nas cincos cidades {media_nas_5_cidades}", 
    f"Media de acidentes de transitos nas cidades com menos de 2000 é {media_2k}"
)

<b>13 Faça um programa que receba o valor de uma dívida e mostre uma tabela com os
seguintes dados: valor da dívida, valor dos juros, quantidade de parcelas e valor da
parcela.<br>
Os juros e a quantidade de parcelas seguem a tabela abaixo:</b><br><br>
Quantidade de Parcelas % de Juros sobre o valor inicial da dívida<br>
- 1 0
- 3 10
- 6 15
- 9 20
- 12 25<br>
Exemplo de saída do programa:<br>
Valor da Dívida Valor dos Juros Quantidade de Parcelas Valor da Parcela<br>
- RS 1.000,00 0 1 RS 1.000,00
- RS 1.100,00 100 3 RS 366,00
- RS 1.150,00 150 6 RS 191,67

In [18]:
import locale
locale.setlocale(locale.LC_ALL, 'pt_BR.UTF-8')

valor_divida = float(input("Digite o valor da divida: "))
print('Valor da Dívida', 'Valor dos Juros', 'Quantidade de Parcelas', 'Valor da Parcela', sep="\t")
[
    print(
        f'{locale.currency(valor_divida * (1+juros/100))}\t{juros}\t\t{parcelas}\t\t\t{locale.currency(valor_divida * (1+juros/100)/parcelas)}'
    )
    for parcelas, juros in [
        [1, 0],
        [3, 10],
        [6, 15],
        [9, 20],
        [12, 25]
    ]
]
print()

Digite o valor da divida: 900
Valor da Dívida	Valor dos Juros	Quantidade de Parcelas	Valor da Parcela
R$ 900,00	0		1			R$ 900,00
R$ 990,00	10		3			R$ 330,00
R$ 1035,00	15		6			R$ 172,50
R$ 1080,00	20		9			R$ 120,00
R$ 1125,00	25		12			R$ 93,75



<b>14 Faça um programa que mostre os n termos da Série a seguir:</b><br><br>
 S = 1/1 + 2/3 + 3/5 + 4/7 + 5/9 + ... + n/m.<br>
Imprima no final a soma da série.

In [20]:
n = int(input("Entre com o total de elementos da série: "))
b = 1
s = ''
soma = 0
for a in range(1, n + 1):
    s += f'{a}/{b}'
    if a != n:
        s += ' + '
    soma += a/b
    a += 1 
    b += 2
    
print(f'S = {s} \nSoma = {soma}')

Entre com o total de elementos da série: 20
S = 1/1 + 2/3 + 3/5 + 4/7 + 5/9 + 6/11 + 7/13 + 8/15 + 9/17 + 10/19 + 11/21 + 12/23 + 13/25 + 14/27 + 15/29 + 16/31 + 17/33 + 18/35 + 19/37 + 20/39 
Soma = 11.239836605182267
